### PR TITLE
fix(eibc): wrong packet written on delayedack  acknowledgment 

### PR DESCRIPTION
**Pull request by danwt**
## Description

----

Closes https://github.com/dymensionxyz/dymension/issues/828

**All** items are required. Please add a note to the item if the item is not applicable and
please add links to any relevant follow-up issues.

PR review checkboxes:

I have...

- [ ]  Added a relevant changelog entry to the `Unreleased` section in `CHANGELOG.md`
- [ ]  Targeted PR against the correct branch
- [ ]  included the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ]  Linked to the GitHub issue with discussion and accepted design
- [ ]  Targets only one GitHub issue
- [ ]  Wrote unit and integration tests
- [ ]  Wrote relevant migration scripts if necessary
- [ ]  All CI checks have passed
- [ ]  Added relevant `godoc` [comments](https://blog.golang.org/godoc-documenting-go-code)
- [ ]  Updated the scripts for local run, e.g genesis_config_commands.sh if the PR changes parameters
- [ ]  Add an issue in the [e2e-tests repo](https://github.com/dymensionxyz/e2e-tests) if necessary

SDK Checklist
- [ ] Import/Export Genesis
- [ ] Registered Invariants
- [ ] Registered Events
- [ ] Updated openapi.yaml
- [ ] No usage of go `map`
- [ ] No usage of `time.Now()`
- [ ] Used fixed point arithmetic and not float arithmetic
- [ ] Avoid panicking in Begin/End block as much as possible
- [ ] No unexpected math Overflow
- [ ] Used `sendCoin` and not `SendCoins`
- [ ] Out-of-block compute is bounded
- [ ] No serialized ID at the end of store keys
- [ ] UInt to byte conversion should use BigEndian

Full security checklist [here](https://www.faulttolerant.xyz/2024-01-16-cosmos-security-1/)

----;

For Reviewer:

- [ ]  Confirmed the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ]  Reviewers assigned
- [ ]  Confirmed all author checklist items have been addressed

---;

After reviewer approval:

- [ ]  In case the PR targets the main branch, PR should not be squash merge in order to keep meaningful git history.
- [ ]  In case the PR targets a release branch, PR must be rebased.

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -220,6 +220,8 @@ require (
 )
 
 replace (
+
+	github.com/cosmos/ibc-go/v6 v6.2.1 => /Users/danwt/Documents/dym/cosmos-ibc-go
```
Comment (omritoptix) [inline go.mod:224]: local replace forgotten ?

File: x/delayedack/keeper/rollapp_packet.go [hunk 2]
```diff
@@ -65,12 +65,15 @@ func (k Keeper) UpdateRollappPacketTransferAddress(
 	}
 	// Set the recipient and sender based on the rollapp packet type
 	recipient, sender := transferPacketData.Receiver, transferPacketData.Sender
+	originalRecipient := rollappPacket.OriginalRecvReceiver // can be empty
```
Comment (omritoptix) [inline x/delayedack/keeper/rollapp_packet.go:68]: why is this line necessary vs just initializing `originalRecipient`? seems to me like the `originalRecipient` will always be empty after that.

File: x/delayedack/rollapp_hooks.go [hunk 3]
```diff
@@ -120,6 +120,18 @@ func (im IBCMiddleware) onRecvPacket(rollappPacket commontypes.RollappPacket, lo
...
 		}
+		/*
+			Here, we do the inverse of what we did when we updated the packet transfer address, when we fulfilled the order
+			to ensure the ack matches what the rollapp expects.
+		*/
+		transferPacketData, err := rollappPacket.GetTransferPacketData()
+		if err != nil {
+			return fmt.Errorf("get transfer packet data: %w", err)
+		}
+		if rollappPacket.OriginalRecvReceiver != "" {
```
Comment (omritoptix) [inline x/delayedack/rollapp_hooks.go:131]: I'd add a comment on why it can be empty here for clarity

**Review by omritoptix — COMMENTED** [review]
overall looks good!
minor comments.

File: x/delayedack/rollapp_hooks.go (hunk 3 above)
Comment (danwt) [reply x/delayedack/rollapp_hooks.go:131]: added

File: x/delayedack/keeper/rollapp_packet.go (hunk 2 above)
Comment (danwt) [reply x/delayedack/keeper/rollapp_packet.go:68]: yeah seems simpler to have empty string 👍  will change

File: go.mod (hunk 1 above)
Comment (danwt) [reply go.mod]: fixed

File: x/delayedack/keeper/finalize.go [hunk 4]
```diff
@@ -94,6 +94,18 @@ func (k Keeper) onRecvPacket(rollappPacket commontypes.RollappPacket, ibc portty
...
 			return
 		}
+		/*
+			Here, we do the inverse of what we did when we updated the packet transfer address, when we fulfilled the order
+			to ensure the ack matches what the rollapp expects.
+		*/
+		transferPacketData, err := rollappPacket.GetTransferPacketData()
+		if err != nil {
+			return fmt.Errorf("get transfer packet data: %w", err)
+		}
```
Comment (mtsitrin) [inline x/delayedack/keeper/finalize.go:104]: all added code can be under this `if`

File: x/delayedack/keeper/rollapp_packet.go [hunk 5]
```diff
@@ -65,12 +65,15 @@ func (k Keeper) UpdateRollappPacketTransferAddress(
...
 	recipient, sender := transferPacketData.Receiver, transferPacketData.Sender
+	var originalRecipient string
 	switch rollappPacket.Type {
 	case commontypes.RollappPacket_ON_RECV:
+		originalRecipient = recipient
 		recipient = address
 	case commontypes.RollappPacket_ON_TIMEOUT:
 		fallthrough
 	case commontypes.RollappPacket_ON_ACK:
+		// TODO: currently we only save the original recipient for RECV, but we could also do it for TIMEOUT and ACK for symmetry if needed
```
Comment (mtsitrin) [inline x/delayedack/keeper/rollapp_packet.go:76]: I would use this `rollappPacket.OriginalRecvReceiver` to keep the original address in all cases (recv/ack/timeout).

it would be used only on `OnRecv`, but it's benefical for queries

File: x/delayedack/keeper/finalize.go [hunk 6]
```diff
@@ -53,7 +55,25 @@ func (k Keeper) finalizeRollappPacket(
...
+						non-eibc: sender will never get the funds back
+						eibc: the fulfiller will never get the funds back, the original target has already been paid
+				If writing it fails and the transfer succeeded, we will never write the ack
+					In this case, the packet is never cleared on the RA
+				If writing it succeeds and the transfer succeeded
+					Happy path
+				If writing it succeeds and the transfer failed
+					We write the errack, the funds will be refunded on the RA
+			            non-eibc: sender will get the funds back
+			            eibc: effective transfer from fulfiller to original target
```
Comment (omritoptix) [inline x/delayedack/keeper/finalize.go:72]:
```suggestion
				We only write the ack if writing it succeeds:
				1. Transfer fails and writing ack fails - In this case, the funds will never be refunded on the RA.
						non-eibc: sender will never get the funds back
						eibc: the fulfiller will never get the funds back, the original target has already been paid
				2. Transfer succeeds and writing ack fails - In this case, the packet is never cleared on the RA.
				3. Transfer succeeds and writing succeeds - happy path
				4. Transfer fails and ack succeeds - we write the err ack and the funds will be refunded on the RA
					 non-eibc: sender will get the funds back
			            eibc: effective transfer from fulfiller to original target
```
